packages/opencensus-core/src/trace/instrumentation/base-plugin.ts

```diff
@@ -58,43 +64,63 @@ export abstract class BasePlugin implements types.Plugin {
    * @param moduleExports nodejs module exports to set as context
    * @param tracer tracer relating to context
    * @param version module version description
+   * @param basedir module absolute path
    */
   // tslint:disable:no-any
-  protected setPluginContext(
+  private setPluginContext(
       moduleExports: any, tracer: modelTypes.Tracer, version: string,
       basedir?: string) {
     this.moduleExports = moduleExports;
     this.tracer = tracer;
     this.version = version;
     this.basedir = basedir;
-    this.logger = tracer.logger || logger.logger();
+    this.logger = tracer.logger;
     this.internalFilesExports = this.loadInternalFiles();
   }
 
 
-  // TODO: review this implementation
-  // From the perspective of an instrumentation module author,
-  // that applyUnpatch is abstract makes it seem like patching is optional,
-  // while unpatching is not. It should be the other way around
-
-  // tslint:disable:no-any
-  applyPatch(
+  /**
+   * Method to apply the instrumentation patch.
+   *
+   * GoF Template Method Pattern - this is the invariant part of the patthern
+   *
+   * @param moduleExports nodejs module exports from the module to patch
+   * @param tracer a tracer instance
+   * @param version version of the current instaled module to patch
+   * @param basedir module absolute path
+   */
+  applyPluginPatch(
+      // tslint:disable:no-any
       moduleExports: any, tracer: modelTypes.Tracer, version: string,
-      basedir?: string): any {
+      basedir: string) {
     this.setPluginContext(moduleExports, tracer, version, basedir);
+    return this.applyPatch();
   }
 
-  abstract applyUnpatch(): void;
+
+  applyPluginUnPatch() {
+    this.applyUnpatch();
+  }
+
+  /**
+   * applyPatch() and applyUnpatch()
+   *
+   * GoF Template Method Pattern - this is the variant part of the pattern
+   * Each plugin should implement his own version
```

Comment:
I wasn't aware that this pattern had a specific name, TIL. However, I don't think this necessary belongs in docs, it may be confusing to users. Not everyone has read GoF.

I would just mention that this method will be called when `applyPluginPatch` is called.

Reply:
Design Patterns are meant to be a kind of universal language, I think it is not confusing at all, it adds more information and makes things clear.  I think our audience is pretty much aware of them.